java-checks/src/main/java/org/sonar/java/checks/RecordInsteadOfClassCheck.java

```diff
@@ -0,0 +1,213 @@
+      .collect(Collectors.toSet());
+
+    if (hasSetter(methods, fieldNames) || !hasGetterForEveryField(methods, fieldNames)) {
+      return;
+    }
+    List<Symbol.MethodSymbol> constructors = classConstructors(methods);
+    if (constructors.size() != 1) {
+      return;
+    }
+    Symbol.MethodSymbol constructor = constructors.get(0);
+    if (hasParameterForEveryField(constructor, fieldNames)) {
+      reportIssue(classTree.simpleName(), String.format("Refactor this class declaration to use 'record %s'.", recordName(classTree, constructor)));
+    }
+  }
+
+  private static List<Symbol.MethodSymbol> classMethods(Symbol.TypeSymbol classSymbol) {
+    return classSymbol
+      .memberSymbols()
+      .stream()
+      .filter(Symbol::isMethodSymbol)
+      .map(Symbol.MethodSymbol.class::cast)
+      .collect(Collectors.toList());
+  }
+
+  private static List<Symbol.VariableSymbol> classFields(Symbol.TypeSymbol classSymbol) {
+    return classSymbol
+      .memberSymbols()
+      .stream()
+      .filter(Symbol::isVariableSymbol)
+      // records can have constant, so discarding them
+      .filter(s -> !isConstant(s))
+      .map(Symbol.VariableSymbol.class::cast)
+      .collect(Collectors.toList());
+  }
+
+  private static List<Symbol.MethodSymbol> classConstructors(List<Symbol.MethodSymbol> methods) {
+    return methods.stream()
+      .filter(m -> "<init>".equals(m.name()))
+      // only explicit constructors
+      .filter(m -> m.declaration() != null)
+      .collect(Collectors.toList());
+  }
+
+  private static boolean hasOnlyPrivateFinalFields(List<Symbol.VariableSymbol> fields) {
+    return fields.stream()
+      .allMatch(RecordInsteadOfClassCheck::isPrivateFinal);
+  }
+
+  private static boolean isConstant(Symbol symbol) {
+    return symbol.isStatic() && symbol.isFinal();
+  }
+
+  private static boolean isPrivateFinal(Symbol symbol) {
+    return symbol.isPrivate() && symbol.isFinal();
+  }
+
+  private static boolean hasSetter(List<Symbol.MethodSymbol> methods, Set<String> fieldNames) {
+    return methods.stream().anyMatch(m -> isSetter(m, fieldNames));
+  }
+
+  private static boolean isSetter(Symbol.MethodSymbol method, Set<String> fieldNames) {
+    String methodName = method.name();
+    return methodName.startsWith("set")
+      && fieldNames.contains(toFieldName(methodName))
+      && method.parameterTypes().size() == 1;
+  }
+
+  private static boolean hasGetterForEveryField(List<Symbol.MethodSymbol> methods, Set<String> fieldNames) {
+    Set<String> gettersForField = methods.stream()
+      .filter(m -> isGetter(m, fieldNames))
+      .map(Symbol::name)
+      .map(RecordInsteadOfClassCheck::toFieldName)
+      .collect(Collectors.toSet());
+    return gettersForField.containsAll(fieldNames);
+  }
+
+  private static boolean isGetter(Symbol.MethodSymbol method, Set<String> fieldNames) {
+    String methodName = method.name();
+    return (methodName.startsWith("get") || methodName.startsWith("is"))
```

Comment:
I feel like we should also support getters whose name is equal to that of the field since a) that's a somewhat common way to name getters and b) that's what the getters will actually be called after converting to a record.

Reply:
👍🏻